addons/crm/static/src/js/forecast/forecast_models.js

```diff
@@ -0,0 +1,163 @@
+     * 
+     * After the read_group, checks the end date of the last displayed group
+     * in order to be able to add the next group with the ForecastColumnQuickCreate widget
+     * @override
+     */
+    async __load(params) {
+        this.baseDomain = params.domain;
+        this.baseContext = params.context;
+        this.forecast_field = params.context.forecast_field;
+        if (!this._isForecast(params)) {
+            return this._super(...arguments);
+        }
+        this.min_groups =
+            params.context.fill_temporal && params.context.fill_temporal.min_groups || undefined;
+        var forecast = this.call('forecast', 'getPeriod', {
+            modelName: params.modelName,
+            fieldName: this.forecast_field,
+            fieldType: params.fields[this.forecast_field].type,
+            granularity: this.granularity,
+            min_groups: this.min_groups,
+            reset: true,
+        });
+        params.domain = forecast.getDomain(this.baseDomain);
+        params.context = forecast.getContext(this.baseContext, false);
+        this.handle = await this._super(...arguments);
+        var lastGroup = this.get(this.handle).data.filter(group => group.value).slice(-1)[0];
+        if (lastGroup) {
+            lastGroup = this.localData[lastGroup.id];
+            forecast.end = moment(lastGroup.range[this.forecast_field].to);
+        }
+        return this.handle;
+    },
+    /**
+     * In order to display the sample data, the context and domain need to be the same as before a reload.
+     * If the user switches to another view (list, pivot, ...) and comes back without adding any record or modifying the
+     * search bar, we still want to display the sample data, but the actual domain and context won't be the same because
+     * load and reload are modifying them.
+     * 
+     * As such, we will only compare the original domain and context (without the use of the ForecastService) to determine
+     * whether the sample data should be enabled or not. We also need to check if the groupby changed as in the original function
+     * @see AbtractModel._haveParamsChanged
+     */
+    __haveParamsChanged(params) {
+        var originalParams = [this.baseContext, this.baseDomain];
+        var currentParams = [params.context, params.domain];
+        if (JSON.stringify(originalParams) !== JSON.stringify(currentParams)) {
+            return true;
+        }
+        if (this.useSampleModel && 'groupBy' in params) {
+            return JSON.stringify(params.groupBy) !== JSON.stringify(this.loadParams.groupedBy);
+        }
+    },
+    /**
+     * Whether the sample data should be displayed (false) or not (true). This had to be computed beforehand,
+     * because `this.__reload` is actually changing the context and the domain, but this function is evaluated
+     * way later @see AbstractModel.reload
+     * @see __haveParamsChanged
+     * @override
+     */
+    _haveParamsChanged() {
```

Comment:
Wait, what ? We have `haveParamsChanged`, `_haveParamsChanged` and `__haveParamsChanged` ? Isn't it a bit the bordel ?

Reply:
That's because what I'm doing here is a total hack xD
Ideally I would write the code which is in __haveParamsChanged in _haveParamsChanged, but the latter is called after `params.domain` and `params.context` already were modified, so it's already too late...
So my current ugly solution is to compute the result that we want beforehand with __haveParamsChanged, store it in haveParamsChanged, and then, when the actual function _haveParamsChanged is called when we want to know if we should display sample data, we ignore the super function and return haveParamsChanged instead...
I'll try to think of a better solution maybe

Reply:
:exploding_head: